diff --git a/pyenet/enet-pyspades.pyx b/pyenet/enet-pyspades.pyx
index 52e81f7..93e9504 100644
--- a/pyenet/enet-pyspades.pyx
+++ b/pyenet/enet-pyspades.pyx
@@ -227,9 +227,6 @@ cdef class Socket:
             &address._enet_address, &buffer, 1)
         return result
 
-    def fileno(self):
-        return self._enet_socket
-
 cdef class Address:
     """
     Address (str address, int port)
@@ -252,7 +249,7 @@ cdef class Address:
     def __init__(self, host, port):
         if host is not None:
             # Convert the hostname to a byte string if needed
-            self.host = host if isinstance(host, bytes) else bytes(host, "ascii")
+            self.host = host
         else:
             self.host = None
         self.port = port
@@ -925,6 +922,7 @@ cdef class Host:
         the host and its peers. The timeout is in milliseconds.
         """
 
+        cdef int result
         if self._enet_host:
             event = Event()
             result = enet_host_service(
@@ -932,6 +930,8 @@ cdef class Host:
 
             if result < 0:
                 raise IOError("Servicing error - probably disconnected.")
+            elif result == 0:
+                return None
             else:
                 return event
 
